catkin_tools/execution/stages.py

```diff
@@ -0,0 +1,127 @@
+
+import os
+
+try:
+    from shlex import quote as cmd_quote
+except ImportError:
+    from pipes import quote as cmd_quote
+
+from .io import IOBufferLogger
+from .io import IOBufferProtocol
+
+
+class Stage(object):
+
+    """A description of one of the serially-executed stages of a Job.
+
+    Like Jobs, Stages are stateless, and simply describe what needs to be done
+    and how to do it.
+    """
+
+    def __init__(self, label, logger_factory=IOBufferProtocol.factory, occupy_job=True, repro=None):
+        self.label = str(label)
+        self.logger_factory = logger_factory
+        self.occupy_job = occupy_job
+        self.repro = repro
+
+    def get_repro(self, verb, jid):
+        """Get a string which should reproduce this stage outside of the catkin command."""
+        return self.repro
+
+
+class CommandStage(Stage):
+
+    """Job stage that describes a system command.
+
+    :param label: The label for the stage
+    :param command: A list of strings composing a system command
+    :param protocol: A protocol class to use for this stage
+
+    :parma cmd: The command to run
+    :param cwd: The directory in which to run the command (default: os.getcwd())
+    :param env: The environment variable overrides. These update the loaded environment (default: {})
+    :param shell: Whether to run the command in "shell" mode, (if True, cmd should be a single string) (default: False)
+    :param emulate_tty: Support colored output (default: True)
+    :param stderr_to_stdout: Pipe stderr to stdout (default: False)
+
+    :param occupy_job: Whether this stage should wait for a worker from the job server (default: True)
+    :param logger_factory: The factory to use to construct a logger (default: IOBufferProtocol.factory)
+
+    """
+
+    def __init__(
+            self,
+            label,
+            cmd,
+            cwd=os.getcwd(),
+            env={},
```

Comment:
This is dangerous (mutable default arguments http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments) please use `None` and do something like `env or {}` where you use it. I know it isn't a problem now, but someone could easily over look this if they update the function in the future.